backend/pkg/modules/internal/differ/spec_diffs.go

```diff
@@ -0,0 +1,285 @@
+	var providedDiff *_spec.APIDiff
+	var err error
+
+	log.Infof("Got new event notification. event=%+v", event)
+
+	apiEvent := event.APIEvent
+	specKey := _speculator.GetSpecKey(apiEvent.HostSpecName, strconv.Itoa(int(apiEvent.DestinationPort)))
+	speculatorAccessor := p.accessor.GetSpeculatorAccessor()
+
+	if !speculatorAccessor.HasProvidedSpec(specKey) && !speculatorAccessor.HasApprovedSpec(specKey) {
+		log.Debugf("No diffs to calculate")
+		return
+	}
+
+	speculatorTelemetry := speculatorutils.ConvertModelsToSpeculatorTelemetry(event.Telemetry)
+
+	reconstructedDiffType := models.DiffTypeNODIFF
+	providedDiffType := models.DiffTypeNODIFF
+	if speculatorAccessor.HasProvidedSpec(specKey) {
+		// calculate diffs base on the event
+		providedDiff, err = speculatorAccessor.DiffTelemetry(speculatorTelemetry, _spec.DiffSourceProvided)
+		if err != nil {
+			log.Errorf("Failed to diff telemetry against provided spec: %v", err)
+			return
+		}
+		if err := setAPIEventProvidedDiff(apiEvent, providedDiff); err != nil {
+			log.Errorf("Failed to set api event provided diff: %v", err)
+			return
+		}
+		providedDiffType = convertToModelsDiffType(providedDiff.Type)
+	}
+	if speculatorAccessor.HasApprovedSpec(specKey) {
+		// calculate diffs base on the event
+		reconstructedDiff, err = speculatorAccessor.DiffTelemetry(speculatorTelemetry, _spec.DiffSourceReconstructed)
+		if err != nil {
+			log.Errorf("Failed to diff telemetry against approved spec: %v", err)
+			return
+		}
+		if err := setAPIEventReconstructedDiff(apiEvent, reconstructedDiff); err != nil {
+			log.Errorf("Failed to set api event reconstructed diff: %v", err)
+			return
+		}
+		reconstructedDiffType = convertToModelsDiffType(reconstructedDiff.Type)
+	}
+
+	apiEvent.SpecDiffType = getHighestPrioritySpecDiffType(providedDiffType, reconstructedDiffType)
+
+	// save api event with diffs in db
+	if err := p.accessor.UpdateAPIEvent(ctx, apiEvent); err != nil {
+		log.Errorf("Failed to update api event: %v", err)
+		return
+	}
+
+	p.addDiffToSend(apiEvent)
+}
+
+func (p *differ) addDiffToSend(event *database.APIEvent) {
+	if event.SpecDiffType == models.DiffTypeNODIFF {
+		return
+	}
+	if p.getTotalEvents() > diffsSendThreshold {
+		log.Warnf("Diff events threshold reached (%v), ignoring event", diffsSendThreshold)
+		return
+	}
+
+	var newSpec string
+	var oldSpec string
+	var hash diffHash
+	// TODO which should get priority?
+	if event.HasReconstructedSpecDiff {
+		newSpec = event.NewReconstructedSpec
+		oldSpec = event.OldReconstructedSpec
+	} else if event.HasProvidedSpecDiff {
+		newSpec = event.NewProvidedSpec
+		oldSpec = event.OldProvidedSpec
+	}
+
+	hash = sha256.Sum256([]byte(newSpec + oldSpec))
+	t := time.Time(event.Time)
+
+	p.Lock()
+	defer p.Unlock()
+	if len(p.apiIDToDiffs[event.APIInfoID]) == 0 {
+		p.apiIDToDiffs[event.APIInfoID] = make(map[diffHash]global.Diff)
+	}
+	if _, ok := p.apiIDToDiffs[event.APIInfoID][hash]; !ok {
+		p.increaseTotalDiffs()
```

Comment:
you have a deadlock here since you lock p again inside the function

Reply:
you right. thanks